src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp

```diff
@@ -0,0 +1,176 @@
+                  sqrt(rho_zero_times_h + square(background_mag_field));
+  fluid_speed_ = -perturbation_size * alfven_speed_ / background_mag_field;
+}
+
+void AlfvenWave::pup(PUP::er& p) noexcept {
+  p | wavenumber_;
+  p | pressure_;
+  p | rest_mass_density_;
+  p | adiabatic_exponent_;
+  p | background_mag_field_;
+  p | perturbation_size_;
+  p | alfven_speed_;
+  p | fluid_speed_;
+}
+
+template <typename DataType>
+DataType AlfvenWave::k_dot_x_minus_vt(const tnsr::I<DataType, 3>& x,
+                                      const double t) const noexcept {
+  auto result = make_with_value<DataType>(x, -wavenumber_ * alfven_speed_ * t);
+  result += wavenumber_ * x.get(2);
+  return result;
+}
+
+template <typename DataType>
+tuples::tagged_tuple_from_typelist<AlfvenWave::variables_t<DataType>>
+AlfvenWave::variables(const tnsr::I<DataType, 3>& x, const double t,
+                      AlfvenWave::variables_t<DataType> /*meta*/) const
+    noexcept {
+  // Explicitly set all variables to zero:
+  auto result = make_with_value<
+      tuples::tagged_tuple_from_typelist<AlfvenWave::variables_t<DataType>>>(
+      x, 0.0);
+
+  const DataType phase = k_dot_x_minus_vt(x, t);
+  get(get<hydro::Tags::RestMassDensity<DataType>>(result)) = rest_mass_density_;
+  get(get<hydro::Tags::SpecificInternalEnergy<DataType>>(result)) =
+      pressure_ / ((adiabatic_exponent_ - 1.0) * rest_mass_density_);
+  get<hydro::Tags::SpatialVelocity<DataType, 3, Frame::Inertial>>(result).get(
+      0) = fluid_speed_ * cos(phase);
+  get<hydro::Tags::SpatialVelocity<DataType, 3, Frame::Inertial>>(result).get(
+      1) = fluid_speed_ * sin(phase);
+  get<hydro::Tags::SpatialVelocity<DataType, 3, Frame::Inertial>>(result).get(
+      2) = 0.0;
+
+  get<hydro::Tags::MagneticField<DataType, 3, Frame::Inertial>>(result).get(0) =
+      perturbation_size_ * cos(phase);
+  get<hydro::Tags::MagneticField<DataType, 3, Frame::Inertial>>(result).get(1) =
+      perturbation_size_ * sin(phase);
+  get<hydro::Tags::MagneticField<DataType, 3, Frame::Inertial>>(result).get(2) =
+      background_mag_field_;
+
+  get(get<hydro::Tags::Pressure<DataType>>(result)) = pressure_;
+  return result;
+}
+
+template <typename DataType>
+tuples::tagged_tuple_from_typelist<AlfvenWave::dt_variables_t<DataType>>
+AlfvenWave::variables(const tnsr::I<DataType, 3>& x, const double t,
+                      AlfvenWave::dt_variables_t<DataType> /*meta*/) const
+    noexcept {
+  // Explicitly set all variables to zero:
+  auto result = make_with_value<
+      tuples::tagged_tuple_from_typelist<AlfvenWave::dt_variables_t<DataType>>>(
+      x, 0.0);
+
+  const DataType phase = k_dot_x_minus_vt(x, t);
+  get(get<Tags::dt<hydro::Tags::SpecificInternalEnergy<DataType>>>(result)) =
+      -pressure_ /
+      ((adiabatic_exponent_ - 1.0) *
+       square(1.0 + perturbation_size_ * sin(phase))) *
+      get(get<Tags::dt<hydro::Tags::RestMassDensity<DataType>>>(result));
```

Comment:
I don't believe this expression can possibly be true.  Isn't it zero?  In fact you say this later.  😕

Reply:
How can your test pass?

Reply:
The only non-zero quantities should be the magnetic field and the velocity field, which vary sinusoidally - this quantity ends up being zero so I shouldn't be setting it at all - thanks for catching this!